### PR TITLE
Updated syntax for easier initialization

### PR DESCRIPTION
**Pull request by thomaswmorris**
This is an attempt to make the overhead to construct the agent less verbose. Having the degrees of freedom and tasks as dictionary also gives us more freedom in the backend, e.g. to add/remove degrees of freedom while sequentially optimizing the beamline. 

Please give suggestions that would improve the syntax.

### CODE REVIEW

File: bloptools/bayesian/__init__.py [hunk 1]
```diff
@@ -284,132 +230,317 @@ def initialize(
...
+                        if dof["kind"] == "active"
+                        else dof["device"].read()[dof["device"].name]["value"]
+                        for dof in self._subset_dofs(mode="on")
+                    ]
+                )
+            ]
+        ).swapaxes(0, -1)
+
+    @property
+    def _acqf_bounds(self):
```
Comment (mrakitin) [inline bloptools/bayesian/__init__.py:390]: `_acqf_bounds` -> `_acq_func_bounds`

File: bloptools/bayesian/__init__.py [hunk 2]
```diff
@@ -284,132 +230,317 @@ def initialize(
...
+            model = task["model"]
+            botorch.fit.fit_gpytorch_mll(gpytorch.mlls.ExactMarginalLogLikelihood(model.likelihood, model), **kwargs)
+        botorch.fit.fit_gpytorch_mll(
+            gpytorch.mlls.ExactMarginalLogLikelihood(self.classifier.likelihood, self.classifier), **kwargs
+        )
         if self.verbose:
             print(f"trained models in {ttime.monotonic() - t0:.02f} seconds")
 
+    @property
+    def acqf_info(self):
```
Comment (mrakitin) [inline bloptools/bayesian/__init__.py:499]: `acqf_info` -> `acq_func_info`

**Review by mrakitin — COMMENTED** [review]
Looks good overall. I've posted a few suggestion to address the things we discussed.

File: bloptools/bayesian/__init__.py [hunk 3]
```diff
@@ -284,132 +230,317 @@ def initialize(
...
+            if "transform" in task.keys():
+                if task["transform"] == "log":
+                    valid &= df.loc[:, name] > 0
+                    df.loc[valid, name] = np.log(df.loc[valid, name])
+                    df.loc[~valid, name] = np.nan
+
+            if task["kind"] == "minimize":
+                df.loc[valid, name] *= -1
+        return df
+
```
[the remark this answers is not in the document]
Comment (thomaswmorris) [reply bloptools/bayesian/__init__.py:332]: I think it should always work with pandas, as long as you make sure everything is mapped to a bool

File: docs/source/usage.rst [hunk 4]
```diff
@@ -2,8 +2,32 @@
...
 
     import bloptools
+
+    dofs = [
+        {"device": some_motor, "limits": (-0.5, 0.5), "kind": "active"},
+        {"device": another_motor, "limits": (-0.5, 0.5), "kind": "active"},
+    ]
+
+    tasks = [
+        {"key": "flux", "kind": "maximize", "transform": "log"}
```
[the remark this answers is not in the document]
Comment (mrakitin) [reply docs/source/usage.rst]: Ah, I don't this it's relevant to `tasks`, the active/passive are for `dofs` only.

File: docs/source/usage.rst [hunk 5]
```diff
@@ -2,8 +2,32 @@
 Usage
 =====
 
-Start by importing bloptools.
+Working in the Bluesky environment, we need to pass four ingredients to the Bayesian agent:
+
+* ``dofs``: A list of degrees of freedom for the agent.
+* ``dets`` (Optional): A list of detectors to be triggered during acquisition.
```
[the remark this answers is not in the document]
Comment (thomaswmorris) [reply docs/source/usage.rst]: I added it to the usage, though it should be optional and not passed most of the time

File: bloptools/bayesian/__init__.py [hunk 6]
```diff
@@ -33,40 +33,97 @@ def default_acquisition_plan(dofs, inputs, dets):
...
     },
 }
 
+TASK_TRANSFORMS = {"log": lambda x: np.log(x)}
+
+
+def _validate_and_prepare_dofs(dofs):
+    for dof in dofs:
+        if type(dof) is not dict:
+            raise ValueError("Supplied dofs must be a list of dicts!")
```
[the remark this answers is not in the document]
Comment (thomaswmorris) [reply bloptools/bayesian/__init__.py]: good idea

**Review by mrakitin — APPROVED** [review]
Amazing work 💯!

**Pull request merged**